OAuth2 support

codingquark opened the pull request:
Adds support for 2Factor Authentication (OAuth2).

New dependencies:
[oauth2](https://elpa.gnu.org/packages/oauth2.html)

Includes readme instructions on how to get relevant keys from mastodon instance and setup local variables.

Makes OAuth2 as default authentication mechanism. This is to be discussed and debated, I suppose.

I would also appreciate feedback on improving or changing test cases.

Commit:
Recognise hashtag links and make them link to our own tag browsing functionality instead.

alexjgriffith committed:
Fix missing thread notifications (#176)

* Print "No Thread\!" if there are no descendants or ancestors.

* Fixed errors for Reblog, follow and favourited notifications.

alexjgriffith committed:
Generalizations for notifications (#174)

* Added generalizations needed for adding notifications #37

- Made generalizations to toot and bylines to enable notifications.
- Fixed tl tests to account for the new function arguments.
- fixed whitespace

Commit:
Clean up code to remove compilation warnings.

We used to be clean but have slipped recently.
Let's clean up the code so it's easier to see real code smell warnings in the compilation.

hdurer committed:
Support toggling content warning anywhere in the toot. (#180)

This new functionality is bound to the 'c' key.

To help with this functionality this also changes the 'toot-json and 'toot-id properties.  These are now applied to the whole toot not just the toot's byline.

alexjgriffith committed:
Fixed jk nav issue and replaced tabs with spaces in mastodon-tl.el

alexjgriffith committed:
Adds the notification timeline to mastodon.el closing #34. (#178)

- Can now handle notifications (follow reblog favourite and mention)
- Tests for each of the rendering functions are included
- Added keybinding N to mastodon.el for mastodon-notifications-get
- added mastodon-notifications to ert-helper.el

hdurer committed:
Optionally use auth-source-search for fetching and saving password. (#181)

* Use auth-source-search for fetching and saving password

This gives users the ability to save their password to either the gpg-encrypted ~/.authinfo.gpg or
~/.authinfo so that they don't have to provide username/password each time

* Add a new custom var to decide whether to use the auth-source package or not.

hdurer committed:
Some small tweaks to our keymap handling. (#183)

- Use special mode which sets up a sensible base keymap.
- Define the keymap just once so that users can tweak it as they wish
  without us messing with that on each invocation of mastodon-mode.

alexjgriffith committed:
Toggle read-only when inserting thread contents. (#184)

alexjgriffith committed:
added mastodon-notifications--get to mastodon.el load in. (#185)

hdurer committed:
Make users handles tabstops (#188) Closes #166

* User ids in the body are tab stops.

* Make user handles in byline also links to profile.

alexjgriffith committed:
Add user profiles closes #165 (#186)

* Add user profiles

- added keybinding U for opening user profiles
- (also added vimish keybindings 'h' and 'l' as aliased for tab and stab in mastodon-mode to compliment jk nav)
- autoloads 'mastodon-profile--get-next-author

alexjgriffith committed:
Adjusted mastodon-tl.el to open user profiles in emacs. (#189)

* Adjusted mastodon-tl.el to open user profiles in emacs.

This also fixes an issue in user profiles where new statuses were insert at the above rather than below the user profile header.

hdurer committed:
Give mastodon-media--inline-images a saner interface. (#191)

Instead of making it search the whole buffer every time to find images to load, give it a range where this work should be done.
We then call this immediately after inserting a single status, notification, ...

There should be no big noticible difference - images might load a tiny bit sooner although I doubt you can see that.
This should be more efficient on large buffers although Alex didn't notice any problems when testing streamed buffers.
We should still do it as it make things easier to understand.  I was always worried about these global operations.

hdurer committed:
Remove compiler warnings in mastodon-notifications.el (#190)

hdurer committed:
More profile work (#193)

* Add an alternative approach to user profile opening.

This way asks the user in the minibuffer for the handle and offering completion for all user handles in the current status but allowing the user to also enter any other handle to browse whichever account they wish.

This also cleans up some compiler warnings about profile code.

* Create a new minor mode for mastodon profile pages.  There we override the 'f' and 'F' keys to show following and followers respectively.

Those pages now look very similar to the regular profile page (with a header).

hdurer committed:
Tiny fixes to `mastodon-inspect--dump-json-in-buffer` (#195)

- Erase buffer before adding new contents
- Don't globally set `print-level` and `print-length`; just temporarily let them to the locally desired values.

hdurer committed:
Fixes to boosting/favoriting. (#194)

I had issues that occasionally I couldn't unfavorite / unboost a toot.
Once that was fixed it annoyed me that the cursor would often jump on those actions (jumping to the byline with no need).

Commit:
Fix the boosting/favouriting of statuses in profiles.

(Favouriting itself is still hard since we have clobbered the 'f' keybinding, but if you bind it to something else or invoke it via `M-x mastodon-toot--toggle-favourite` then at least it works.)
This also changes the regular boosting/favoriting behavior.  Before we would boost/fav a boost or fav instead of the actual toot that was boosted/faved.  With this change we always boost/fav the base toot.

To do this we now keep a second toot id (with the base toot) in the byline's properities.  (For regular statuses 'toot-id and 'base-toot-id will be identical.)

Commit:
Improvements on posting toots.

- Visual feedback in the buffer of:
  - Number of characters
  - Content warning flag
  - Posting visibility
- Can't post an empty toot.
- Changing visibility to something other than "public".

Still missing:
 - Attaching media
 - Enabling the NSWF-flag toggling.

jdenen committed:
Clean up whitespace

jdenen committed:
Rebind and document some keybindings

hdurer committed:
Fix boosting and faving in the notifications buffer.

Now that we calculate the base toot id, we can ensure that it is unset in notifications that do not represent a toot.

jdenen committed:
Bump to version 0.8.0

jdenen committed:
Update README

juergenhoetzel committed:
Use the email address of the current user as default (#203)

* Use the email address of the current user as default

jdormit committed:
Add ability to change the visibility

Commit:
Make default visibility customizable; add documentation.

jdenen committed:
Make the test less fragile

Commit:
[WIP] Basic flow to fetch OAuth2 token upon logging in

Commit:
[FIX] Customize group value menu shows the options

Commit:
GET, POST requests over OAuth2!

Commit:
Bad variable names

Commit:
Ugly code

Commit:
Custom vars to help with OAuth2 lib

Commit:
No need to define auth-sources, make sure client id/secret exist

Commit:
Variable names were too generic

Commit:
[WIP] Readme notes for 2FA

Commit:
[FIX] Invalid check string for plain token generation

Commit:
Previous COND suffices for the steps to be taken

Commit:
Update tests to have auth mechanism specified

Commit:
ReadMe instructions to get client id and secret

Commit:
Basic OAuth2 tests

hdurer commented on README.org:
```diff
@@ -69,6 +65,40 @@ There is an option to have your user credentials (email address and password) sa
+2. =mastodon-oauth2-client-id=
+3. =mastodon-oauth2-client-secret=
+
+**** Do the following to obtain OAuth2 credentials:
```
Isn't this exactly what `mastodon-client--register` does?  Why does the user have to do that manually now?

codingquark replied:
I see, missed that one. I shall refactor.

hdurer commented on lisp/mastodon-auth.el:
```diff
@@ -60,10 +81,11 @@ if you are happy with unencryped storage use e.g. \"~/authinfo\"."
+           (string= "plain" mastodon-auth-mechanism))
       (mastodon-auth--generate-token-no-storing-credentials)
-    (mastodon-auth--generate-token-and-store)))
+    (mastodon-oauth2--generate-token-and-store)))
```
Does that mean we can no longer store the credentials when using `plain` mode?
Does that mean it goes through the token generation for each API call when using `plain`?

codingquark replied:
The OAuth2 library stores the credentials in its own file, which means it doesn't affect the way we store the credentials in `plain` mode. The method will check if the auth mechanism is `plain` or `oauth2`, and will call token generation procedure accordingly.

The OAuth2 library uses the existing tokens, refreshes them if needed, or generates new ones if not available. Hence, the calls will go through the procedure, but HTTP APIs will only be called when necessary.

hdurer reviewed (commented):
I had just a quick first look.  Looks generally ok.  I don't understand if this causes functionality regressions for plain mode.

Commit:
Client id, secret are not to be set explicitly by the user

`mastodon-client--register` - which is an existing method - handles
the generation of new application in the user's account, store the ID
and token locally. The simplest entry point to read/fetch the
application details is to call `mastodon-client` because it reads if
the values are locally available, or makes appropriate API calls to
fetch them from the instance; making the life simple by returning an
easily accessible plist.

Thus, I have removed the requirement of user having to create new
application themselves. Now the `mastodon-client` is called while
generating new oauth2 tokens.

Commit:
[FIX] Logic to decide which mechanism to use for token gen

Previously, the logic was WRONG. It would never have stored
credentials. Now it applies proper logic to go through a few
conditions to decide if OAuth2 is to be stored or plain tokens are to
be stored.

Commit:
[FIX] Test exceeding max-lisp-eval-depth

Commit:
[FIX] The test is going to fail, because instance.url not fake!

Commit:
[FIX] package name prefix added

Commit:
[FIX] package name prefix added

Really sorry, I'm high on caffaine, and being too rash.

Commit:
[CI] fixing on v0.11 of oauth2.el

Commit:
[FIX] OMG

hdurer commented on README.org:
```diff
@@ -69,6 +65,22 @@ There is an option to have your user credentials (email address and password) sa
+
+The OAuth2 library uses =plstore= to securely store the credentials.
+However, =mastodon-oauth2-client-secret= needs to be stored in a safe
+place - encryption is recommended.
```
That isn't very actionable advice.  Where is that stored and how does the user ensure it is encrypted?

codingquark replied:
Updated, how is it now?

hdurer commented on README.org:
```diff
@@ -69,6 +65,22 @@ There is an option to have your user credentials (email address and password) sa
+  (setq mastodon-auth-mechanism "oauth2")
+#+END_SRC
+
+The OAuth2 library uses =plstore= to securely store the credentials.
```
Can you add a link somewhere appropriate to the oauth2 library (plus maybe a note which version)?

E.g. I don't use melpa, so I'll need to find the library and manually add it to me config files repo.

codingquark replied:
oauth2.el is available on elpa. Do you think we should link it?

hdurer replied:
I found and use https://github.com/emacsmirror/oauth2.git in my own setup.

hdurer commented on README.org:
```diff
@@ -69,6 +65,22 @@ There is an option to have your user credentials (email address and password) sa
+  (setq mastodon-auth-mechanism "oauth2")
+#+END_SRC
+
+The OAuth2 library uses =plstore= to securely store the credentials.
```
Is that plstore location configurable?  Maybe mention the appropriate var?

codingquark replied:
Added info on steps.

hdurer replied:
Seems the var is `oauth2-token-file`, we could/should mention that explicitly here.

Actually, could we just temporarily around oauth operations let `oauth2-token-file` to be the same value as `mastodon-auth-source-file` and thereby use only one secrets storage?

codingquark replied:
I had used `defvaralias` for a while. But I decided to go against it. Since this is a generic library which can be used as a dependency by other packages, I think aliasing it will not be a good idea.

Is there any alternative approach so that the global env is not spoiled?

hdurer commented on lisp/mastodon-auth.el:
```diff
@@ -108,6 +112,17 @@ Reads and/or stores secres in `MASTODON-AUTH-SOURCE-FILE'."
+(defun mastodon-auth-oauth2--generate-token-and-store ()
+  "Generate OAuth2 token.
+
+Reads and/or stores secres in `OAUTH2-TOKEN-FILE'."
```
s/secres/secrets/

hdurer replied:
Ping!  This still says "secres"

hdurer commented on lisp/mastodon-http.el:
```diff
@@ -83,10 +84,17 @@ Authorization header is included by default unless UNAUTHENTICED-P is non-nil."
 	 (append
 	  (unless unauthenticed-p
-	    `(("Authorization" . ,(concat "Bearer " (mastodon-auth--access-token)))))
+	    `(("Authorization" . ,(concat "Bearer " (if (string= "oauth2" mastodon-auth-mechanism)
```
Is this header needed at all for oauth2 or should the `(string= "oauth2" mastodon-auth-mechanism)` check go into the `unless` above to totally bypass this in the new flow?

codingquark replied:
I have updated post and get both.

Commit:
Better explanation

Commit:
squash! Better explanation

Commit:
Typo

Commit:
Removed unnecessary sections

hdurer commented on README.org:
```diff
@@ -69,6 +65,24 @@ There is an option to have your user credentials (email address and password) sa
+
+**** Set up
+#+BEGIN_SRC emacs-lisp
+  (setq mastodon-auth-mechanism "oauth2")
```
Actually it seem that  `mastodon-auth-mechanism` should be a symbol (that's what you define as options in the defcustom.  So this should be `'oauth2` and not `"oauth2"`.

It would be clearer, if you had consistently use `(eq 'oauth2 mastodon-auth-mechanism)` instead of doing a `(string= ...`.

hdurer replied:
Nit: `'oauth` - no more strings.

hdurer commented on lisp/mastodon-http.el:
```diff
@@ -82,21 +83,31 @@ Authorization header is included by default unless UNAUTHENTICED-P is non-nil."
         (url-request-extra-headers
 	 (append
-	  (unless unauthenticed-p
+	  (unless (and unauthenticed-p
```
As mentioned in the other big comment, I think this should be `or` not `and`.

hdurer commented on lisp/mastodon-http.el:
```diff
@@ -82,21 +83,31 @@ Authorization header is included by default unless UNAUTHENTICED-P is non-nil."
 	  headers)))
     (with-temp-buffer
-      (url-retrieve-synchronously url))))
+      (if (string= "oauth2" mastodon-auth-mechanism)
```
As mentioned in other comment: Don't use `oauth2-url-retrieve-synchronously` when making an unauthenticated call.

codingquark replied:
Instead of this, I am going with the modification in `mastodon-client--register`. This will make sure the settings are changed only when needed and will keep logical clarity intact.

In my point of view, we don't want to use oauth2 to register the client. Which has nothing to do with generic post requests directly. Although it is true that not making calls to `oauth2-url-retrieve-synchronously` is a solution, it is at a place which is too broad.

I hope I am making the reasoning clear.

codingquark replied:
Now that I'm thinking about it, fixing it at the POST call will prevent future bugs too.

codingquark replied:
@hdurer this is regarding the fix being put at two places.

hdurer reviewed (commented):
I really like this... only a few more comments.

Commit:
Link to OAuth2 library being used

Commit:
[FIX] That broke everything

Commit:
Since the mechanism is quote, equality should check for symbol

Commit:
[FIX] Prevent infinite loop while registering client

For more than one accounts, the package would go in an infinite loop
at startup. While registering the client, it makes a POST
request. This POST request requires a TOKEN if the mechanism is 'oauth2.

Commit:
[FIX] Preventing potential infinite loops

Call to oauth2-url-retrieve-synchronously requires a TOKEN to be
present. If the user is unauthenticated, the TOKEN will not be
present. This is going to cause infinite loops.

hdurer commented on lisp/mastodon-auth.el:
```diff
@@ -60,10 +61,13 @@ if you are happy with unencryped storage use e.g. \"~/authinfo\"."
-    (mastodon-auth--generate-token-and-store)))
+  (cond ((and (or (null mastodon-auth-source-file)
+                  (string= "" mastodon-auth-source-file))
+              (string= "plain" mastodon-auth-mechanism))
```
Nit: `(eq 'plain mastodon-auth-mechanism)`

hdurer commented on test/mastodon-auth-tests.el:
```diff
@@ -4,7 +4,8 @@
    (let ((mastodon-auth-source-file "")
-	 (mastodon-instance-url "https://instance.url"))
+	 (mastodon-instance-url "https://instance.url")
+         (mastodon-auth-mechanism "plain"))
```
Nit: `'plain`   (and a few more elsewhere...)

hdurer commented on test/mastodon-oauth2-tests.el:
```diff
@@ -0,0 +1,49 @@
+  :expected-result :failed
+  (with-mock
+   (let ((mastodon-instance-url "https://instance.url")
+         (mastodon-auth-mechanism "oauth2"))
```
Nit: `'oauth2`  (and 2 more below)

hdurer commented on lisp/mastodon-client.el:
```diff
@@ -45,14 +46,15 @@
-     ("website" . "https://github.com/jdenen/mastodon.el"))
-   nil
-   :unauthenticated))
+  (let ((mastodon-auth-mechanism 'plain))
```
Is this now actually still needed?  You also applied the other fix (checking whether the call is unauthenticated), right?

hdurer commented on lisp/mastodon-http.el:
```diff
@@ -82,21 +83,32 @@ Authorization header is included by default unless UNAUTHENTICED-P is non-nil."
 	 (append
-	  (unless unauthenticed-p
+	  (unless (or unauthenticed-p
+                      (eq 'oauth2 mastodon-auth-mechanism))
```
That indentation looks weird.

When you are done, can you run `M-x untabbify` on all touched files to get rid of these pesky tab characters?

hdurer reviewed (commented):
Looking good.
Just some nits left as comments.

I still want to use this code longer than 30s but in general I am really happy about this.
We should come up with some path to get rid of the old auth mechanism - or is there any use for it?

Commit:
Typo

Commit:
No strings anymore

Commit:
No more strings!

Commit:
Untabified

Commit:
No more strings!

Commit:
oauth2 as non-dev dependency too

Commit:
Trying to keep CI happy

Commit:
Revert "Trying to keep CI happy"

This reverts commit b7de30f8b2dfd7ea446bcb6531ee7cff13cc770d.

Commit:
Trying to keep CI happy

hdurer reviewed (approved):
As mentioned before, I am happy with this code. nwo.

Commit:
Env sanity

jdenen committed:
Merge branch 'develop' into oauth2_0
